control-plane-operator/controllers/hostedcontrolplane/olm/catalogs.go

```diff
@@ -118,19 +122,47 @@ func findTagReference(tags []imagev1.TagReference, name string) *imagev1.TagRefe
 	return nil
 }
 
-var CatalogToImage = map[string]string{
-	"certified-operators": "registry.redhat.io/redhat/certified-operator-index:v4.16",
-	"community-operators": "registry.redhat.io/redhat/community-operator-index:v4.16",
-	"redhat-marketplace":  "registry.redhat.io/redhat/redhat-marketplace-index:v4.16",
-	"redhat-operators":    "registry.redhat.io/redhat/redhat-operator-index:v4.16",
+func GetCatalogImages(ctx context.Context, hcp hyperv1.HostedControlPlane, pullSecret []byte, digestLister registryclient.DigestListerFN) (map[string]string, error) {
+	imageRef := hcp.Spec.ReleaseImage
+	imageConfig, _, _, err := registryclient.GetMetadata(ctx, imageRef, pullSecret)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get image metadata: %w", err)
+	}
+
+	version, err := semver.Parse(imageConfig.Config.Labels["io.openshift.release"])
+	if err != nil {
+		return nil, fmt.Errorf("invalid OpenShift release version format: %s", imageConfig.Config.Labels["io.openshift.release"])
+	}
+
+	operators := map[string]string{
+		"certified-operators": fmt.Sprintf("registry.redhat.io/redhat/certified-operator-index:v%s", version.String()),
+		"community-operators": fmt.Sprintf("registry.redhat.io/redhat/community-operator-index:v%s", version.String()),
+		"redhat-marketplace":  fmt.Sprintf("registry.redhat.io/redhat/redhat-marketplace-index:v%s", version.String()),
+		"redhat-operators":    fmt.Sprintf("registry.redhat.io/redhat/redhat-operator-index:v%s", version.String()),
+	}
+
+	for i := 0; i < 4; i++ {
+		_, err = digestLister(ctx, operators["certified-operators"], pullSecret)
+		if err != nil {
```

Comment:
were lowering the version by 1 until we find one that's available from the last 4 supported versions this is so that right after branching this doesn't fail on us when the latest catalog images arent available yet. last 4 version might be overkill.

Reply:
right, but why we do that when `digestLister()` returns an error?

Reply:
I guess the way this for loop is written is not clear, how about this

```go
func GetCatalogImages(ctx context.Context, hcp hyperv1.HostedControlPlane, pullSecret []byte, digestLister registryclient.DigestListerFN) (map[string]string, error) {
	imageRef := hcp.Spec.ReleaseImage
	imageConfig, _, _, err := registryclient.GetMetadata(ctx, imageRef, pullSecret)
	if err != nil {
		return nil, fmt.Errorf("failed to get image metadata: %w", err)
	}

	version, err := semver.Parse(imageConfig.Config.Labels["io.openshift.release"])
	if err != nil {
		return nil, fmt.Errorf("invalid OpenShift release version format: %s", imageConfig.Config.Labels["io.openshift.release"])
	}

	imageFound := false
	for i := 0; i < 4; i++ {
		imageRef := fmt.Sprintf("registry.redhat.io/redhat/certified-operator-index:v%s", version.String())
		_, err = digestLister(ctx, imageRef, pullSecret)
		if err == nil {
			// found an available image.
			imageFound = true
			break
		}
		if !strings.Contains(err.Error(), "manifest unknown") {
			return nil, err
		}

		// keep lowering the minor version until we find an available image.
		version.Minor--
	}

	if !imageFound {
		return nil, fmt.Errorf("couldn't find an image")
	}

	operators := map[string]string{
		"certified-operators": fmt.Sprintf("registry.redhat.io/redhat/certified-operator-index:v%s", version.String()),
		"community-operators": fmt.Sprintf("registry.redhat.io/redhat/community-operator-index:v%s", version.String()),
		"redhat-marketplace":  fmt.Sprintf("registry.redhat.io/redhat/redhat-marketplace-index:v%s", version.String()),
		"redhat-operators":    fmt.Sprintf("registry.redhat.io/redhat/redhat-operator-index:v%s", version.String()),
	}
	return operators, nil
}
```

Reply:
manifestUnknow error means that the manifest doesnt exist anything else points to other errors that are unexpected.

Reply:
you need to break on `err == nil`, otherwise you will always run the loop 4 times, even if you found a valid image immediately. Making the break/exist condition of a loop explicit is always better.


Also we need to return a readable error if we don't find an image withing the supported versions, rather thann returning whatever generic error `digestLister()` returns.